common/src/seg_tree.c

```diff
@@ -217,6 +217,40 @@ int seg_tree_add(struct seg_tree* seg_tree, unsigned long start,
     return 0;
 }
 
+/* search tree for entry that overlaps with given start/end
+ * offsets, return first overlapping entry if found, NULL otherwise,
+ * assumes caller has lock on tree */
+struct seg_tree_node* seg_tree_find(
+    struct seg_tree* seg_tree,
+    unsigned long start,
+    unsigned long end)
+{
+    /* Create a range of just our starting byte offset */
+    struct seg_tree_node* node = seg_tree_node_alloc(start, start, 0);
+    if (!node) {
+        return NULL;
+    }
+
+    /* search tree for either a range that overlaps with
+     * the target range (starting byte), or otherwise the
+     * node for the next biggest starting byte */
+    struct seg_tree_node* next = RB_NFIND(inttree, &seg_tree->head, node);
+
+    free(node);
+
+    /* we may have found a node that doesn't include our starting
+     * byte offset, but it would be the range with the lowest
+     * starting offset after the target starting offset, check whether
+     * this overlaps our end offset */
```

Comment:
This is super clever :+1:

Reply:
Thanks.  Only thought to come up with that because I had a bug in my code before figuring this out.